### PR TITLE
Modified genCellLibrary.tcl to output all lib cell properties to cellLibrary.xml file

### PR DESCRIPTION
**Pull request by nelsobe**
[cellLibrary.txt](https://github.com/byuccl/tincr/files/484034/cellLibrary.txt)

Decided to just dump all properties for now.  May decide we only want a subset later.  

Comments?

### CODE REVIEW

**Comment by ttown523** [comment]
Looks good. Can you attach the new version of the cellLibrary.xml so I can see what the properties look like when printed?

**Review by bradselw — APPROVED** [review]
Looks good to me

**Comment by nelsobe** [comment]
Attached a sample output file.

I made no attempt to parse the types of the properties or anything else.  They all show up in Vivado as string properties, I just output them exactly as Vivado shows them

**Comment by ttown523** [comment]
After looking at the cellLibrary.xml file, I have a few comments. 

I don't think it's correct to simply dump out each of the properties with a value. Many properties are related, and so should be consolidated into one. For example, lines 28-50 in the file you posted look like: 

```
        <libcellproperty>
          <name>CONFIG.JTAG_CHAIN</name>
          <value></value>
        </libcellproperty>
        <libcellproperty>
          <name>CONFIG.JTAG_CHAIN.DEFAULT</name>
          <value>1</value>
        </libcellproperty>
        <libcellproperty>
          <name>CONFIG.JTAG_CHAIN.MAX</name>
          <value>4</value>
        </libcellproperty>
        <libcellproperty>
          <name>CONFIG.JTAG_CHAIN.MIN</name>
          <value>1</value>
        </libcellproperty>
        <libcellproperty>
          <name>CONFIG.JTAG_CHAIN.TYPE</name>
          <value>int</value>
        </libcellproperty>
        <libcellproperty>
          <name>CONFIG.JTAG_CHAIN.VALUES</name>
          <value>1, 2, 3, 4</value>
        </libcellproperty>
```

This list of libcellproperties is really one property with the name, default, type, max, min, and possible values. So, it should probably look something like: 

```
<libcellproperty>
  <name>CONFIG.JTAG_CHAIN</name>
  <type>int</type>
  <default_value>1</default_value>
  <max>4</max>
  <min>1</min>
  <possible>
    <value>1</value>
    <value>2</value>
    <value>3</value>
    <value>4</value>
  </possible>
</libcellproperty>
```

The possible values could also be a range for such properties as an INIT string instead of a list of discrete values. This would make parsing these lib cell properties into RapidSmith much easier. 

Also, some properties we can probably just filter out completely. For example, the following properties are probably not needed in RapidSmith at all. The class property because we already know it's a lib cell, the name because we already have the lib cell name, and the Xilinx_Library because we would never use that information. There are several other properties like this.

```
        <libcellproperty>
          <name>CLASS</name>
          <value>lib_cell</value>
        </libcellproperty>
        <libcellproperty>
          <name>NAME</name>
          <value>BSCANE2</value>
        </libcellproperty>
        <libcellproperty>
          <name>XILINX_LIBRARY</name>
          <value>UNISIM</value>
        </libcellproperty>
```

Now that I look more carefully at the properties of library cells, the only ones that I think are useful to keep are any CONFIG properties and the PRIMITIVE_GROUP property. The CONFIG properties are settable, which would allow a RapidSmith users to set these properties to the value they want. The PRIMITIVE_GROUP property would be useful for identification/cell grouping (i.e. looking for all LUTs in the design without having to look at the name of the library cell every time). We could make this a read-only property.

Thoughts?

**Comment by nelsobe** [comment]
Regarding filtering out some properties:  I originally filtered out and only printed the CONFIG.\* properties.  After thinking about it I figured that having them all may have some value so I removed the filter.  It could be switched back (it was in two different commits).  My reason for including them all is that (a) they don’t cause much problem (unless we realize they double the memory footprint of the cell library data structure) and (b) there may be some value in having some of them available that we currently don’t know about – we don’t want to have to keep going back and modifying the filter.   Thoughts?

Regarding grouping related properties, I thought of doing that as well.  My reason for not doing it was simply it wasn't clear if the time to do it was worth the value added.  For me, including them as I did is as much as about documenting the possible properties that could be used as anything…

**Comment by ttown523** [comment]
I don't think any of the lib cell properties other than the ones I proposed have any value in RapidSmith. Either we already have access to the property, or they are Xilinx related things that will never make sense to put in RapidSmith. Do you have an example of a property that may be useful some day, because I don't see one. If for some reason we do want to add another property to RapidSmith, I don't mind adding these after the fact.

Also, there are several problems in having them. For one, they are included in every cell in the cell library, and so this takes up a significant amount a space (this is the smallest problem). Another problem is that we are duplicating data that already exists. For example, the NAME and NUM_PINS is already in RapidSmith, there is no reason to include them again. The final, and most important reason, is that several of these properties have an empty value field (for example PRIMITIVE_SUBGROUP). If there is no value, and the property is not configurable, what does this mean in RapidSmith? The user can set the property to whatever they want? These properties are read-only in Vivado.

Your reason for not grouping properties is not clear to me. For example, "CONFIG.JTAG_CHAIN.DEFAULT" is not a property of the cell. It is the default value of the CONFIG.JTAG_CHAIN property. It doesn't make sense to have this as a separate cell property. The whole idea was to create properties on lib cells in RapidSmith that contained the name of the property, the default value, and other relevant information of the property. This is significantly easier in RapidSmith if we used the way I proposed above, and I don't think it should take very long to implement.  

Also, I think it's best to not respond directly via email to these conversation threads. It not only includes your message, but the message you are replying to which clogs the pull request page up.

**Comment by nelsobe** [comment]
[cellLibrary.txt](https://github.com/byuccl/tincr/files/486509/cellLibrary.txt)

Modified the code to group properties when writing them out.

**Comment by bradselw** [comment]
I'm not exactly sure how RS consumes this file, but it may be helpful to specify whether or not the property is read-only in Vivaldi.

**Comment by nelsobe** [comment]
Brad brings up a good point.  
In Tcl, they are all marked as read-only since they are attached to library cells.
The working assumption is that the CONFIG.\* properties are writeable on cell instances.  The PRIMITIVE_GROUP property is not.

File: tincr/io/library/genCellLibrary.tcl [hunk 1]
```diff
@@ -384,20 +384,41 @@ proc write_port_xml { fo lc dict } {
...
+    }
+    
+    if { [llength $lst] > 0 } {
+	puts $fo "      <libcellproperties>"
+	foreach h $lst {
 	    puts $fo "        <libcellproperty>"
-	    puts $fo "          <name>$p</name>"
-	    puts $fo "          <value>[get_property $p $lc]</value>"
+	    puts $fo "          <name>$h</name>"
+	    if { $h == "PRIMITIVE_GROUP" } {
```
Comment (ttown523) [inline tincr/io/library/genCellLibrary.tcl]: I would change this to `if { [string first "CONFIG" $h] == -1}` to print out property values of any thing that is not a config. If we use this approach and decide that we want to add more properties in the future, we won't have to keep adding to this code. Also, piggybacking on Brad's comment, we should probably mark these with the `<readonly/>` tag.

File: tincr/io/library/genCellLibrary.tcl [hunk 2]
```diff
@@ -384,20 +384,41 @@ proc write_port_xml { fo lc dict } {
...
+		puts $fo "          <value>[get_property $h $lc]</value>"
+	    }
+	    
+	    foreach p $pl {
+		if { [llength [split $p "."]] == 3 && [string first $h $p] == 0 } {
+		    set tag [lindex [split $p "."] 2]
```
Comment (ttown523) [inline tincr/io/library/genCellLibrary.tcl:417]: To be consistent in the XML, should we convert the tag to lower case? 
`[string tolower $tag]`

**Review by ttown523 — COMMENTED** [review]
For the most part, it looks good. But, I think that the variable names need to be more descriptive. This helps readability and anyone else who may look at the code in the future.

**Comment by nelsobe** [comment]
I thought the same thing about lower case (they are ugly).  But, there may be value in them being exactly what Vivado outputs (and possibly what they require on input), that is why I output them unchanged.

I don't understand the previous comment about  using: if { [string first "CONFIG" $h] == -1} instead of if { $h == "PRIMITIVE_GROUP" } ...  There are many non-CONFIG properties - the suggested code will catch them all.  I thought the only non-CONFIG one we wanted to match was PRIMITIVE_GROUP, thus the test for that string.  Please clarify.

**Comment by ttown523** [comment]
At the part of the code that I added the comment, the properties to print are already in the "lst" list. In other words, all the properties that we want to include in the XML are in "lst." Therefore we know that if a property is not a CONFIG property, it is a read-only property.

**Comment by nelsobe** [comment]
Thanks for all the good suggestions!

I have lower-cased the sub-property names.

I now understand where Thomas was pointing in the code (I was looking in the wrong place) and have made the change to be more general.

I have also added the readonly tag to all but the CONFIG properties.

**Pull request merged**